calc/vxlan_resolver.go

```diff
@@ -0,0 +1,453 @@
+				// If we've already handled this node, there's nothing to do. Deduplicate.
+				logCxt.Debug("Skipping duplicate node IP update")
+				return
+			}
+
+			// We've already sent a VTEP for this node, and the node's IP address has changed.
+			// We need to revoke it and any routes before sending any updates.
+			routes := c.nodeNameToSentRoutes[nodeName]
+			c.withdrawRoutes(routes)
+			c.sendVTEPRemove(nodeName)
+		}
+
+		// Try sending a VTEP update. If we do, this will trigger a kick of any
+		// pending routes which might now be ready to send.
+		c.sendVTEPUpdate(nodeName)
+	} else {
+		// Withdraw any routes which target this VTEP, followed by the VTEP itself.
+		delete(c.nodeNameToIPAddr, nodeName)
+		routes := c.nodeNameToSentRoutes[nodeName]
+		c.withdrawRoutes(routes)
+		c.sendVTEPRemove(nodeName)
+	}
+	return
+}
+
+// OnHostConfigUpdate gets called whenever a node's host config changes. We only care about
+// VXLAN tunnel address updates. On an add/update, we need to check if there are VTEPs or routes which
+// are now valid, and trigger programming of them to the data plane. On a delete, we need to withdraw any
+// routes and VTEPs associated with the node.
+func (c *VXLANResolver) OnHostConfigUpdate(update api.Update) (_ bool) {
+	switch update.Key.(model.HostConfigKey).Name {
+	case "IPv4VXLANTunnelAddr":
+		nodeName := update.Key.(model.HostConfigKey).Hostname
+		logCxt := logrus.WithField("node", nodeName)
+		if update.Value != nil {
+			// Update for a VXLAN tunnel address.
+			newIP := update.Value.(string)
+			currIP := c.nodeNameToVXLANTunnelAddr[nodeName]
+			c.nodeNameToVXLANTunnelAddr[nodeName] = newIP
+			logCxt = logCxt.WithFields(logrus.Fields{"newIP": newIP, "currIP": currIP})
+			if c.sentVTEPs[nodeName] {
+				if currIP == newIP {
+					// If we've already handled this node, there's nothing to do. Deduplicate.
+					logCxt.Debug("Skipping duplicate tunnel addr update")
+					return
+				}
+
+				// We've already sent a VTEP for this node, and the node's IP address has changed.
+				// We need to revoke it and any routes before sending any updates.
+				routes := c.nodeNameToSentRoutes[nodeName]
+				c.withdrawRoutes(routes)
+				c.sendVTEPRemove(nodeName)
+			}
+
+			// Try sending a VTEP update. If we do, this will trigger a kick of any
+			// pending routes which might now be ready to send.
+			c.sendVTEPUpdate(nodeName)
+		} else {
+			// Withdraw any routes which target this VTEP, followed by the VTEP itself.
+			delete(c.nodeNameToVXLANTunnelAddr, nodeName)
+			routes := c.nodeNameToSentRoutes[nodeName]
+			c.withdrawRoutes(routes)
+			c.sendVTEPRemove(nodeName)
+		}
+	}
+	return
+}
+
+// OnPoolUpdate gets called whenever an IP pool changes. If a new VXLAN pool is added, kick
+// pending routes to see if any should now be programmed. If a VXLAN pool is removed, then
+// find any routes which need to be withdrawn and remove them. Pool CIDRs cannot be modified,
+// so we do not need to handle that here.
```

Comment:
One man's delete/create can be another man's update.  Imagine this scenario:

- IP pool "my-pool" has CIDR 10.0.0.0/16
- Felix disconnected from datastore
- "my-pool" gets deleted
- New "my-pool" gets created with different CIDR
- Felix resyncs with datastore.  Syncer sends an update, not a delete/add.

Granted, that's pretty unlikely but if we let that type of problem stack up in the codebase, our robustness overall will drop.  Suggest you just sanity check that the CIDR is the same and if not, handle as a delete and re-add.